File: 백준/Platinum/10217. KCM Travel/KCM Travel.py
import sys
input = sys.stdin.readline


def Main():
    for tc in range(int(input())):
        N,M,K = map(int,input().split())
        arr = [[]for _ in range(N+1)]
        for i in range(K):
            u,v,c,d = map(int,input().split())
            arr[u].append((v,c,d))
        dp = [[float("inf")]*(M+1) for _ in range(N+1)]
        dp[1][0] = 0
        for cost in range(M+1):
            for node in range(1, N+1):
                if dp[node][cost] != float('inf'):
                    for next, nextC,nextD in arr[node]:
                        curC = cost + nextC
                        curD = dp[node][cost]+nextD
                        if curC > M:
                            continue
                        if dp[next][curC] > curD:
                            dp[next][curC] = curD
        ans = min(dp[N])
        if ans == float('inf'):print("Poor KCM")
        else:print(ans)
Main()

File: 백준/Platinum/10217. KCM Travel/KCM Travel.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static class Node {
        int v,c, d;

        Node(int v, int c, int d) {
            this.v = v;
            this.c = c;
            this.d = d;
        }
    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());
        for (int tc = 0; tc < T; tc++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int N = Integer.parseInt(st.nextToken());
            int M = Integer.parseInt(st.nextToken());
            int K = Integer.parseInt(st.nextToken());
            ArrayList<Node>[] nodes = new ArrayList[N+1];
            for (int i = 0; i <= N; i++) {
                nodes[i] = new ArrayList<>();
            }
            for (int i = 0; i < K; i++) {
                st = new StringTokenizer(br.readLine());
                int u = Integer.parseInt(st.nextToken());
                int v = Integer.parseInt(st.nextToken());
                int c = Integer.parseInt(st.nextToken());
                int d = Integer.parseInt(st.nextToken());
                nodes[u].add(new Node(v,c,d));
            }
            int[][] dp = new int[N + 1][M + 1];
            for (int i = 0; i <= N; i++) {
                Arrays.fill(dp[i], Integer.MAX_VALUE);
            }
            dp[1][0] = 0;
            for (int cost = 0; cost <= M; cost++) {
                for (int now = 1; now < N; now++) {
                    if (dp[now][cost] != Integer.MAX_VALUE) {
                        for (Node node : nodes[now]) {
                            int curC = cost + node.c;
                            int curD = dp[now][cost] + node.d;
                            if (curC > M) {
                                continue;
                            }
                            if (dp[node.v][curC] > curD) {
                                dp[node.v][curC] = curD;
                            }
                        }
                    }
                }
            }
            int ans = Integer.MAX_VALUE;
            for (int cost = 0; cost <= M; cost++) {
                ans = Math.min(ans, dp[N][cost]);
            }
            if (ans == Integer.MAX_VALUE) {
                System.out.println("Poor KCM");
            } else {
                System.out.println(ans);
            }
        }
    }
}